Refactorizations of the ReDoS libraries

erik-krogh opened the pull request:
Many things are happening in this PR, all of which are outlined below.  

Commit-by-commit review is highly recommended.   
All the tests should pass on every commit.  

## ~~1: Disable parts of QL-for-QL~~

~~The consistency and `ql/dead-code` queries are very noisy until we have proper support for parameterized modules. 
So I've disabled those.~~ 
No longer needed now that https://github.com/github/codeql/pull/9575 has been merged.   

## 2: Introduce a `ReDoSPruning` parameterized module 

The `class Config extends string` pattern used previously was ugly.  
This is just a straight-forward conversion of that pattern into a parameterized module. 

## 3: Implement a more efficient algorithm for computing strings from chains

Based on @hvitved's work: https://github.com/github/codeql/pull/8589 

But moved into a parameterized module in order to share more code between exponential and superlinear ReDoS.  
It does the same thing as @hvitved's implementation, but in a slightly different way (to make a pretty signature module).  

## 4: add further normalization of char classses 

Inspired by @esbena: https://github.com/github/codeql/pull/4847#discussion_r547244758 

Now `\d` and `[\d]` and `[0-9]` will literally be represented with a single value.  
(an arbitrary one of them is chosen as the canonical representative).  

The extra normalization has a real, but limited, impact.  
The largest impact is on the size of the `SuperlinearBackTracking::getAThreewayIntersect` predicate, the size of that predicate drops by ~15% on large JS benchmarks. 

I also drive-by fixed a bug related to case-normalization.  
(The bug didn't affect anything before the changes in this PR).  

## 5: make a parameterized module out of the RegexpMatching implementation

I didn't like the configuration pattern used previously, which was also why I hadn't made a real library out of it.  
It can be better with parameterized modules, so I did that. 

## 6: Rename `ReDoSUtil` to `NfaUtils`, and rename the `performance` folder to `regexp`.  

The `ReDoSUtil` file was used for more than just ReDoS: it's a general library for constructing and reasoning about NFAs.   
And the `performance` folder was also a bad name, so I renamed it to `regexp`.  

## 7: Improve performance.   

Lots of states in the `process()` predicate could be eliminated.   
Because we know that if a regexp ever "reaches" the accept-all state, then it can never be rejected, and thus we can skip them early.      

I've also ported [this exponential-redos fix from Shack](https://github.com/github/codeql/pull/9381) to the superlinear library. 

----  

## Evaluations. 

All look good, and mostly show a significant speedup (except for Python, which has neutral performance).  
[JavaScript](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/redosNormal-js/reports), [Ruby](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/redosNormal-ruby/reports), [Java](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/redosNormal-java/reports), [Python](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/redosNormal-python/reports).   

---- 

Previous version of this PR, which got filled up with noise: https://github.com/github/codeql/pull/8522

aschackmull commented:
Ping @ginsbach : Parameterised modules in here.

ginsbach commented:
Thanks for pinging. The use of parameterised modules here looks good to me, no concerns!

erik-krogh commented:
I've rebased on the latest `main` now that QL-for-QL supports parameterized modules: https://github.com/github/codeql/pull/9575   

I've also re-enabled the QL-for-QL consistency- and dead-code-query in this PR.

joefarebrother commented on java/ql/lib/semmle/code/java/security/regexp/NfaUtils.qll:
```diff
@@ -0,0 +1,1322 @@
+  }
+
+  /** Gets a string representing all the chars matched by `c` */
+  private string getMormalizationString(CharacterClass c) {
```
Should this be `getNormalizationString`?

erik-krogh replied:
Yes it should!

erik-krogh commented:
I did a rebase on `main` to resolve the conflict with https://github.com/github/codeql/pull/9618.

asgerf commented on javascript/ql/lib/semmle/javascript/security/performance/RegexpMatching.qll:
```diff
@@ -0,0 +1,175 @@
+}
+
+/**
+ * Holds if it should be tested whether `reg` matches `str`.
```
```suggestion
 * Holds if it should be tested whether `root` matches `str`.
````

asgerf commented on java/ql/lib/semmle/code/java/security/performance/ReDoSUtil.qll:
```diff
@@ -1240,3 +1240,82 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
+/**
+ * A module that describes a tree where each node has one or more accosiated characters.
+ * The root node has no accosiated character.
+ * This module is a signature used in `Concretizer`.
```
```suggestion
 * A module that describes a tree where each node has one or more associated characters.
 * The root node has no associated character.
 * This module is a signature used in `Concretizer`.
```

asgerf replied:
If I understand correctly, this data structure is commonly known as a Trie.

asgerf commented on java/ql/lib/semmle/code/java/security/performance/ReDoSUtil.qll:
```diff
@@ -1240,3 +1240,82 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
+
+  /**
+   * Holds if `n` is at the end of a tree. I.e. a node that should have a result in the `Concretizer` module.
+   * A leaf can still have children.
```
```suggestion
   * Such a node can still have children.
```

asgerf commented on java/ql/lib/semmle/code/java/security/performance/ReDoSUtil.qll:
```diff
@@ -1240,3 +1240,82 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
+    exists(Node prev | result = nodeDist(prev) + 1 | prev = getPrev(n))
+  }
+
+  /** Holds if `n` is part of a chain that we want to compute a string for. */
```
```suggestion
  /** Gets an ancestor of `end`. */
```

asgerf commented on java/ql/lib/semmle/code/java/security/performance/ReDoSUtil.qll:
```diff
@@ -1240,3 +1240,82 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
+  private predicate isRoot(Node n) { not exists(getPrev(n)) }
+
+  /** Gets the distance from a root to `n`. */
+  private int nodeDist(Node n) {
```
`nodeDepth`?

asgerf commented on java/ql/lib/semmle/code/java/security/performance/ReDoSUtil.qll:
```diff
@@ -1240,3 +1240,82 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
+  }
+
+  pragma[noinline]
+  private string getPrefixChar(Node n, int i) {
```
```suggestion
  /**
   * Gets the `i`th character on the path from `n` back to the root (i.e. counting backwards).
   */
  pragma[noinline]
  private string getPrefixChar(Node n, int i) {
```

asgerf replied:
The qldoc I suggested above is based on what the predicate appears to do now.

But if we set `i = nodeDist(prev)` this would just be the `i`th character (1-based) in the string, right? (1-based because the root has no character). That seems like a more natural thing to work with than this difference thing.

Then we'd just have to remove `desc` from the concat ordering below.

asgerf commented on java/ql/lib/semmle/code/java/security/regexp/NfaUtils.qll:
```diff
@@ -890,6 +890,7 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
    */
   private predicate isReDoSCandidate(State state, string pump) {
     isCandidate(state, pump) and
+    not epsilonSucc*(state) = AcceptAnySuffix(_) and // pruning early - these can never get stuck in a rejecting state.
```
Is there a reason not to use the `acceptsAnySuffix()` helper predicate here and in `reachesOnlyRejectableSuffixes`?

erik-krogh replied:
No there isn't. I just forgot that I had it.

nickrolfe commented on python/ql/lib/change-notes/2022-05-25-redos-refac.md:
```diff
@@ -0,0 +1,5 @@
+---
+category: deprecated
+---
+* The utility files previous in the `semmle.python.security.performance` package has been moved to the `semmle.python.security.regexp` package.  
```
```suggestion
* The utility files previously in the `semmle.python.security.performance` package have been moved to the `semmle.python.security.regexp` package.  
```

asgerf reviewed (commented):
Mostly LGTM with a few nits to address

I didn't review code that appears to have been mostly moved, and for now I've only skimmed the commit which "simplifies recursion" as it wasn't obvious from the description here what's really going on.

erik-krogh commented:
> If I understood correctly, NfaUtils is still an old-style Pyramaterised module. Any plans to change that? 

Yes. But doing that requires features that we don't have yet (nested modules, and better support for classes in parameterized modules).   
The entire API for the tree-structure of an regular-expressions needs to become a signature.

erik-krogh commented:
I did a merge with `main` to fix a merge conflict.  

@github/codeql-python can I get a review?

yoff commented on python/ql/lib/semmle/python/security/regexp/RegexpMatching.qll:
```diff
@@ -0,0 +1,175 @@
+   * Starts with an accepting state as found by `getAState` and searches backwards
+   * to the start state through the reachable states (as found by `getAState`).
+   *
+   * This predicate holds the invariant that the result state can be reached with `i` steps from a start state,
```
```suggestion
   * This predicate satisfies the invariant that the result state can be reached with `i` steps from a start state,
```

yoff commented on python/ql/lib/semmle/python/security/regexp/RegexpMatching.qll:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  /**
+   * Holds if `reg` matches `str`, where `str` is either in the `test` or `testWithGroups` predicate.
```
```suggestion
   * Holds if `reg` matches `str`, where `str` is in the `isCandidate` predicate.
```

yoff commented on python/ql/lib/semmle/python/security/regexp/NfaUtils.qll:
```diff
@@ -0,0 +1,1323 @@
+  private predicate isRelevant(Node n) {
+    isARelevantEnd(n)
+    or
+    exists(Node prev | isRelevant(prev) | n = getPrev(prev))
```
```suggestion
    exists(Node succ | isRelevant(succ) | n = getPrev(succ))
```

yoff commented on python/ql/lib/semmle/python/security/regexp/NfaUtils.qll:
```diff
@@ -0,0 +1,1323 @@
+    exists(Node prev | result = nodeDepth(prev) + 1 | prev = getPrev(n))
+  }
+
+  /** Gets an ancestor of `end`, where `end` is a node that should have a result in `concretize`. */
+  private Node getANodeInLongChain(Node end) {
```
Could this just be called `getAnAncestor`? And could it be written via `prev*`?

erik-krogh replied:
Maybe it can, I don't remember if I wrote the recursion manually for a reason 🤷  

I'll do another evaluation to check.

yoff commented on python/ql/lib/semmle/python/security/regexp/NfaUtils.qll:
```diff
@@ -0,0 +1,1323 @@
+  /** Gets the `i`th character on the path from the root to `n`. */
+  pragma[noinline]
+  private string getPrefixChar(Node n, int i) {
+    exists(Node prev |
```
Maybe rename `prev` to `ancestor` (or an abbreviation thereof)

yoff commented on python/ql/lib/semmle/python/security/regexp/NfaUtils.qll:
```diff
@@ -0,0 +1,1323 @@
+    predicate reachesOnlyRejectableSuffixes(State fork, string w) {
+      isReDoSCandidate(fork, w) and
+      forex(State next | next = process(fork, w, w.length() - 1) | isLikelyRejectable(next)) and
+      not epsilonSucc*(getProcessPrevious(fork, _, w)) = AcceptAnySuffix(_) // we stop `process(..)` early if we can, check here if it happened.
```
I think you can rewrite it like this
```suggestion
      not getProcessPrevious(fork, _, w) = acceptsAnySuffix() // we stop `process(..)` early if we can, check here if it happened.
```

yoff reviewed (changes requested):
I really love this work (although it could arguably have been in several PRs).

The requested changes are all minor.

erik-krogh commented on python/ql/lib/semmle/python/security/regexp/NfaUtils.qll:
```diff
@@ -0,0 +1,1323 @@
+     */
+    pragma[noinline]
+    private predicate isLikelyRejectable(State s) {
+      s = stateInPumpableRegexp() and
```
Yep. 
![Screenshot 2022-08-12 at 19 53 56](https://user-images.githubusercontent.com/54990334/184415789-f8cd0fd7-6300-4ea1-93ac-78c15d7e079f.png) 

I can only extend database types inside a parameterised module.  
And that's not feasible here.

erik-krogh commented on python/ql/lib/semmle/python/security/BadTagFilterQuery.qll:
```diff
@@ -200,12 +29,25 @@ class HtmlMatchingRegExp extends RegexpMatching::MatchedRegExp {
+  )
 }
 
-/** DEPRECATED: Alias for HtmlMatchingRegExp */
-deprecated class HTMLMatchingRegExp = HtmlMatchingRegExp;
```
No, I'll reintroduce it.  
It's in a `Query.qll` so it's likely fine, but I'll get it back anyway.

erik-krogh commented:
@yoff I think I've addressed your review comments, and I also addressed some additional documentation issues identified by QL-for-QL.  

The remaining QL-for-QL alerts are false positives that will be fixed by https://github.com/github/codeql/pull/9884 

I've started new evaluations for [JavaScript](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-js/reports),  [Python](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-py/reports), [Java](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-java/reports), and [Ruby](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-rb/reports).

yoff commented:
Yep, looks great :-)

yoff reviewed (approved):
Lgtm

erik-krogh commented:
> I've started new evaluations for [JavaScript](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-js/reports), [Python](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-py/reports), [Java](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-java/reports), and [Ruby](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/refacReDoS-rb/reports).

The evaluations are all looking good.
